shared/common-adapters/avatar.js

```diff
@@ -0,0 +1,211 @@
+  '48': {bottom: 0, left: 32, position: 'absolute'},
+  '64': {bottom: 0, left: 45, position: 'absolute'},
+  '80': {bottom: 0, left: 57, position: 'absolute'},
+}
+
+class Avatar extends Component<void, Props, State> {
+  state: State
+  _onURLLoaded: ?(username: string, urlMap: ?{[key: string]: string}) => void
+
+  constructor (props: Props) {
+    super(props)
+    if (props.url && props.username) {
+      console.warn('Recieved both url and username to avatar!')
+    }
+
+    this.state = {
+      url: null,
+    }
+  }
+
+  componentWillMount () {
+    if (this.props.url) {
+      this.setState({url: this.props.url})
+      // Just let it load the url, prefer this over username
+    } else if (this.props.username) {
+      this._loadUsername(this.props.username)
+    } else { // Just show the no avatar state
+      this.setState({url: this._urlMapsToUrl(null)})
+    }
+  }
+
+  componentDidMount () {
+    this._onURLLoaded = (username: string, urlMap: ?URLMap) => {
+      // Still looking at the same username?
+      if (this.props.username === username) {
+        this.setState({url: this._urlMapsToUrl(urlMap)})
+      }
+    }
+  }
+
+  componentWillUnmount () {
+    // Don't let the callback do anything if we've unmounted
+    this._onURLLoaded = null
+  }
+
+  _noAvatar () {
+    return iconTypeToImgSet(placeHolders[String(this.props.size)])
+  }
+
+  // Figure out which 1x, 2x images to use for this size
+  _urlMapsToUrl (urlMap: ?URLMap) {
+    if (!urlMap) {
+      return null
+    }
+
+    const low = urlMap['40'] || urlMap['200'] || urlMap['360']
+    const medium = urlMap['200'] || urlMap['360'] || urlMap['40']
+    const high = urlMap['360'] || urlMap['200'] || urlMap['40']
+
+    let imgs = []
+    switch (this.props.size) {
+      case 176:
+      case 112: // fallthrough
+        imgs = [medium, high]
+        break
+      case 80:
+      case 64: // fallthrough
+      case 48: // fallthrough
+        imgs = [medium, medium]
+        break
+      case 40:
+      case 32: // fallthrough
+        imgs = [low, medium]
+        break
+      case 24:
+      case 16: // fallthrough
+        imgs = [low, low]
+        break
+    }
+
+    const m = imgs.reduce((map, img, idx) => {
+      if (img) {
+        map[String(idx + 1)] = img
+      }
+      return map
+    }, {})
+
+    return urlsToImgSet(m) || this._noAvatar()
+  }
+
+  _loadUsername (username: string) {
+    const urlMap = _avatarToURL ? _avatarToURL(username) : null
+    const url = this._urlMapsToUrl(urlMap)
+    this.setState({url})
+
+    if (!urlMap && _loadAvatarToURL) { // Have to load it
+      _loadAvatarToURL(username, (username: string, urlMap: ?URLMap) => {
+        this._onURLLoaded && this._onURLLoaded(username, urlMap)
+      })
+    }
+  }
+
+  componentWillReceiveProps (nextProps: Props) {
+    if (nextProps.url && nextProps.username) {
+      console.warn('Recieved both url and username to avatar!')
+      return
```

Comment:
if both are passed url is used, we can't mutate props (unless we want an inner class or something just to do this (or recompose))